src/bvh/optimization.rs

```diff
@@ -0,0 +1,1039 @@
+            if let &BVHNode::Node { parent_index, child_l_index, child_r_index, .. } =
+                &self.nodes[node_index] {
+                (parent_index, child_l_index, child_r_index)
+            } else {
+                unreachable!()
+            };
+
+        // Recalculate AABBs for the children since at least one of them changed.
+        // Don't update the AABBs in the node yet because they're still subject to change
+        // during potential upcoming rotations.
+        let child_l_aabb = self.nodes[child_l_index].get_node_aabb(shapes);
+        let child_r_aabb = self.nodes[child_r_index].get_node_aabb(shapes);
+
+        // Contains the surface area that would result from applying the currently favored rotation.
+        // The rotation with the smallest SA will be applied in the end.
+        // The value is calculated by `child_l_aabb.surface_area() + child_r_aabb.surface_area()`.
+        let mut best_surface_area = child_l_aabb.surface_area() + child_r_aabb.surface_area();
+
+        #[derive(Debug, Copy, Clone)]
+        struct NodeData {
+            index: usize,
+            aabb: AABB,
+        }
+
+        // Get the grandchildren's NodeData
+        let get_children_node_data = |node: &BVHNode| -> Option<(NodeData, NodeData)> {
+            match node {
+                &BVHNode::Node { child_l_index, child_l_aabb, child_r_index, child_r_aabb, .. } => {
+                    Some((NodeData {
+                              index: child_l_index,
+                              aabb: child_l_aabb,
+                          },
+                          NodeData {
+                              index: child_r_index,
+                              aabb: child_r_aabb,
+                          }))
+                }
+                &BVHNode::Leaf { .. } => None,
+            }
+        };
+
+        // Get indices and AABBs of all grandchildren
+        let left_children_nodes = get_children_node_data(&self.nodes[child_l_index]);
+        let right_children_nodes = get_children_node_data(&self.nodes[child_r_index]);
+
+        // Stores the Rotation that would result in the surface area best_surface_area,
+        // thus being the favored rotation that will be executed after considering all rotations.
+        // Consider all the possible rotations, choose the one with the lowest SAH cost.
+        let mut best_rotation: Option<(usize, usize)> = None;
+        {
+            let mut consider_rotation = |new_rotation: (usize, usize), sa: f32| {
+                info!("        SA {} vs. current SA {}.", sa, best_surface_area);
+                if sa < best_surface_area {
+                    best_surface_area = sa;
+                    best_rotation = Some(new_rotation);
+                }
+            };
+
+            // Child to grandchild rotations
+            if let Some((child_rl, child_rr)) = right_children_nodes {
+                {
+                    let sa = child_rl.aabb.surface_area() +
```

Comment:
Each line is different, and it brings no merit to define two highly specialized functions, which would wrap this. Actually you can't wrap this in a more concise and selfexplanatory way.